CustomControlLibrary/CurrencyConverterControl.cs

```diff
@@ -0,0 +1,509 @@
+                    return;
+                }
+                _currencyToValue = value;
+                var bind = new Binding(nameof(CurrencyToAmount))
+                {
+                    Source = this,
+                    Mode = BindingMode.TwoWay,
+                    UpdateSourceTrigger = UpdateSourceTrigger.PropertyChanged,
+                    Converter = new NumberToStringConvertor(),
+                    Delay = 350
+                };
+                _currencyToValue.SetBinding(TextBox.TextProperty, bind);
+            }
+        }
+
+        private void CurrencyFromOnSelectionChanged(object sender, SelectionChangedEventArgs e)
+        {
+            var selectedItem = CurrencyFrom.SelectedItem as CountryInfo;
+            if(selectedItem==null)
+            {
+                return;
+            }
+            SelectedCurrencyIdFrom = selectedItem.CurrencyId;
+        }
+
+        private void CurrencyToOnSelectionChanged(object sender, SelectionChangedEventArgs e)
+        {
+            var selectedItem = CurrencyTo.SelectedItem as CountryInfo;
+            if (selectedItem == null)
+            {
+                return;
+            }
+
+            SelectedCurrencyIdTo = selectedItem.CurrencyId;
+        }
+
+        #endregion
+
+        private bool _preventCycleCalculation;
+        private IExchangeRateProvider _exchangeRateProvider;
+        private CancellationTokenSource _cancellationTokenSource;
+
+        static CurrencyConverterControl()
+        {
+            DefaultStyleKeyProperty.OverrideMetadata(typeof(CurrencyConverterControl), new FrameworkPropertyMetadata(typeof(CurrencyConverterControl)));
+        }
+
+        public CurrencyConverterControl()
+        {
+            InitializeModel();
+            Unloaded += OnUnloaded;
+        }
+
+        private void OnUnloaded(object sender, RoutedEventArgs e)
+        {
+            Unloaded -= OnUnloaded;
+
+            FinalizeUI();
+            FinalizeModel();
+        }
+
+        public override void OnApplyTemplate()
+        {
+            InitializeUI();
+            TryInvalidateExchangeRate();
+            base.OnApplyTemplate();
+        }
+
+        private void InitializeModel()
+        {
+            var resource = Application.GetResourceStream(new Uri("/CustomControlLibrary;component/assets/countries.json", UriKind.Relative));
+            DataContractJsonSerializer ser = new DataContractJsonSerializer(typeof(List<CountryInfo>));
+            var countries = (ser.ReadObject(resource.Stream) as List<CountryInfo>);
+ 
+            SetValue(CounriesItemsKey, countries);
+        }
+
+        private void FinalizeModel()
+        {
+            CounriesItems.Clear();
+            
+            _cancellationTokenSource?.Cancel();
+            _cancellationTokenSource = null;
+        }
+
+        private bool _uiInited;
+        private void InitializeUI()
+        {
+            CurrencyFrom = (ComboBox)GetTemplateChild("PART_CurrencyFrom");
```

Comment:
It might not be clear that CurrencyFrom is a control.
For these control, we usually name them as _elementCurrencyFrom.
The events are all attached in InitializeUI and detached in FinalizeUI. I think attach and detach when assign value to property like yours is also ok. But should make sure the property/variable name is clear enough for other devs to know they are control type.

Reply:
ok.